generic/zookeepertcl.c

```diff
@@ -1973,6 +1899,49 @@ zootcl_delete_subcommand(Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[], ZO
 	return zootcl_set_tcl_return_code (interp, status);
 }
 
+/*
+ *----------------------------------------------------------------------
+ *
+ * zootcl_destroy_subcommand --
+ *
+ *      implement the "destroy" method of a zookeeper tcl command
+ *      object
+ *
+ * Results:
+ *      Always returns TCL_OK whether or not anything in this function
+ *		succeeds
+ *
+ *
+ *----------------------------------------------------------------------
+ */
+int
+zootcl_destroy_subcommand(Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[], ZOOAPI zhandle_t *zh, zootcl_objectClientData *zo)
```

Comment:
Why is this not using the existing object deletion handler `zootcl_zookeeperObjectDelete`?

Reply:
If the body of `zootcl_destroy_subcommand` is turned into:

```
        zootcl_zookeeperObjectDelete ((ClientData)zo);
        return 1;
```

Then calling `zk destroy` causes `assert (zo->zookeeper_object_magic == ZOOKEEPER_OBJECT_MAGIC);` in `zootcl_zookeeperObjectObjCmd` to fail.  This can be seen in the `create_ephemeral_node` unit test in `tests/api.test`.

Also, in a tclsh, with the above code for `zootcl_destroy_subcommand`, running `zk init ...`, `zk destroy`, `zk init ...` and `zk destroy` ends up causing a segfault.  Or, running `zk init ...`, `zk destroy` and then `zk destroy` again causes the above mentioned assertion error.  Using `zootcl_zookeeperObjectDelete`, in short, does not properly clean up the Zookeeper handle object the way the `zootcl_destroy_subcommand` does.

Reply:
Maybe have both call a common `zootcl_cleanup_handle()` to do the heavy lifting? It's a complex enough chunk of code it doesn't need to be in two places.

Reply:
Yeah, good point.  I'll go ahead and do that.

Reply:
My mistake: in the above code snippet I needed to return `TCL_OK` rather than `1` and then everything works out.

Reply:
In addition, I also realized that it's possible to make an async call with a Zookeeper object and then destroy it before the async callback is invoked.  When this happens, the program will abort because an `assert` will fail.  To deal with this, I've also added a call to `Tcl_DeleteEvents` in `zootcl_zookeeperObjectObjCmd`.  Please let me know what you think about that approach.